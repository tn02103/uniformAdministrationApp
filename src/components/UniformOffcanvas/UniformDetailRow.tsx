import { updateUniformItem } from "@/dal/uniform/item/_index";
import { useUniformGenerationListByType, useUniformTypeList } from "@/dataFetcher/uniformAdmin";
import { useI18n } from "@/lib/locales/client";
import { SAFormHandler } from "@/lib/SAFormHandler";
import { getUniformSizelist } from "@/lib/uniformHelper";
import { useBreakpoint } from "@/lib/useBreakpoint";
import { Uniform, UniformType } from "@/types/globalUniformTypes";
import { getUniformFormSchema, UniformFormType } from "@/zod/uniform";
import { zodResolver } from "@hookform/resolvers/zod";
import { useEffect, useMemo } from "react";
import { Button, Col, Row } from "react-bootstrap";
import { FormProvider, useForm } from "react-hook-form";
import { SelectFormField } from "../fields/SelectFormField";
import { TextareaFormField } from "../fields/TextareaFormField";
import { ToggleFormField } from "../fields/ToggleFormField";
import { useGlobalData } from "../globalDataProvider";

export type UniformDetailRowProps = {
    uniform: Uniform;
    uniformType: UniformType;
    editable: boolean;
    setEditable: (editable: boolean) => void;
    onSave: () => void;
}
const getUniformFormData = (uniform: Uniform): UniformFormType => {
    return {
        id: uniform.id,
        number: uniform.number,
        generation: uniform.generation?.id,
        size: uniform.size?.id,
        comment: uniform.comment || "",
        active: uniform.active || false,
    }
}

export const UniformDetailRow = ({ uniform, uniformType, editable, setEditable, onSave }: UniformDetailRowProps) => {
    // HOOKS
    const t = useI18n();
    const form = useForm<UniformFormType>({
        mode: 'onTouched',
        defaultValues: getUniformFormData(uniform),
        resolver: zodResolver(getUniformFormSchema(uniformType.usingGenerations, uniformType.usingSizes)),
    });
    const { reset, handleSubmit, watch, setValue, setError } = form;

    // DATA Fetcher
    const { generationList } = useUniformGenerationListByType(uniform.type.id);
    const { sizelists } = useGlobalData();
    const { typeList } = useUniformTypeList();
    const selectedGeneration = watch('generation');
    const selectedSizeId = watch('size');
    const { match: isSizeXS } = useBreakpoint("xs", 'eq');

    // DATA PROCESSING
    const sizelist = useMemo(() => {
        if (!typeList) return null;
        return getUniformSizelist({
            type: uniformType,
            generationId: selectedGeneration,
            sizelists: sizelists,
        });
    }, [uniform.type.id, selectedGeneration]);
    const sizeOptions = useMemo(() => {
        if (!editable) {
            if (uniform.size)
                return [{ value: uniform.size?.id, label: uniform.size?.name }];
            return []
        }
        return sizelist?.uniformSizes.map(s => ({ value: s.id, label: s.name })) || [];
    }, [sizelist, editable, uniform.size]);
    const generationOptions = useMemo(() => {
        return generationList?.map(g => ({ value: g.id, label: g.name })) || [];
    }, [generationList]);

    // USE EFFECTS
    useEffect(() => {
        if (!editable) {
            reset(getUniformFormData(uniform));
        }
    }, [uniform]);

    useEffect(() => {
        if (editable && selectedSizeId && sizeOptions) {
            const selectedSize = sizeOptions.find(s => s.value === selectedSizeId);
            if (!selectedSize) {
                setValue("size", "");
            }
        }
    }, [selectedSizeId, sizeOptions, setValue, editable]);

    // DATA HANDLERS
    const handleSave = async (data: UniformFormType) => {
        await SAFormHandler<Uniform, UniformFormType>(
            updateUniformItem(data),
            setError,
            (result) => {
                setEditable(false);
                reset(getUniformFormData(result));
                onSave();
            },
            t('common.error.actions.save')
        );
    }

    return (
        <FormProvider {...form}>
            <form noValidate autoComplete="off" onSubmit={handleSubmit(handleSave)} className="mb-4">
                <Row className="mb-3 mt-2 justify-content-between">
                    <Col xs={(uniformType.usingGenerations && uniformType.usingSizes) ? 12 : 5} sm={"auto"}>
                        <ToggleFormField<UniformFormType>
                            label={t('common.status')}
                            name={"active"}
                            formName="uniform"
                            disabled={!editable}
                            hideToggle={!editable}
                            toggleText={t(`common.uniform.active.${form.watch('active') ? "true" : "false"}`)}
                        />
                    </Col>
                    {uniformType.usingGenerations &&
                        <Col xs={7} sm={"auto"} style={{ minWidth: (uniformType.usingSizes || isSizeXS) ? undefined : "300px" }}>
                            <SelectFormField<UniformFormType>
                                name="generation"
                                label={t('common.uniform.generation.label', { count: 1 })}
                                formName="uniform"
                                disabled={!editable}
                                plaintext={!editable}
                                options={generationOptions}
                            />
                        </Col>
                    }
                    {uniformType.usingSizes &&
                        <Col xs={5} sm={"auto"} className="p-0" style={{ minWidth: isSizeXS ? undefined : uniformType.usingGenerations ? "100px" : "300px" }}>
                            <SelectFormField<UniformFormType>
                                name="size"
                                label={t('common.uniform.size')}
                                formName="uniform"
                                disabled={!editable}
                                plaintext={!editable}
                                options={sizeOptions}
                            />
                        </Col>
                    }
                    <Col xs={12}>
                        <TextareaFormField<UniformFormType>
                            name="comment"
                            label={t('common.comment')}
                            formName="uniform"
                            disabled={!editable}
                            plaintext={!editable}
                            rows={2}
                        />
                    </Col>
                </Row>
                {editable &&
                    <Row className="justify-content-evenly">
                        <Col xs="auto">
                            <Button
                                variant="outline-secondary"
                                type="button"
                                onClick={() => {
                                    setEditable(false);
                                    reset(getUniformFormData(uniform));
                                }}
                            >
                                {t('common.actions.cancel')}
                            </Button>
                        </Col>
                        <Col xs="auto">
                            <Button
                                variant="outline-primary"
                                type="submit"
                            >
                                {t('common.actions.save')}
                            </Button>
                        </Col>
                    </Row>
                }
            </form>
        </FormProvider>
    )
}
