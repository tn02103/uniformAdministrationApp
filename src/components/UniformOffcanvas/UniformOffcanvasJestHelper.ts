import { Deficiency } from "@/types/deficiencyTypes";
import { generationLists, sizeLists, typeList } from "../../../tests/_jestConfig/staticMockData";
import { AuthRole } from "@/lib/AuthRoles";


// ------------- MOCKS FOR DEFICIENCY -------------
jest.mock('@/dal/inspection/deficiency', () => ({
    createUniformDeficiency: jest.fn(async () => "created successfully"),
    resolveDeficiency: jest.fn(async () => "resolved successfully"),
    updateUniformDeficiency: jest.fn(async () => "updated successfully"),
}));

jest.mock('@/dataFetcher/deficiency', () => ({
    useDeficienciesByUniformId: jest.fn((_, includeResolved) => {
        if (includeResolved) {
            return { deficiencies: mockDeficiencyList };
        } else {
            return { deficiencies: mockDeficiencyList.slice(0, 2) };
        }
    }),
    useDeficiencyTypes: jest.fn(() => ({ deficiencyTypeList: mockDeficiencyTypeList })),
}));

// ------------- MOCKS FOR UNIFORM -------------

jest.mock('@/dataFetcher/uniformAdmin', () => ({
    useUniformGenerationListByType: jest.fn(() => ({
        generationList: generationLists[0]
    })),
    useUniformTypeList: jest.fn(() => ({
        typeList
    })),
}));
jest.mock("@/dal/uniform/item/_index", () => ({
    updateUniformItem: jest.fn(() => Promise.resolve('Saved item')),
    deleteUniformItem: jest.fn(() => Promise.resolve('Deleted item')),
    getUniformItemHistory: jest.fn(() => Promise.resolve([])), // Mock for history
    getUniformItemCountByType: jest.fn(() => Promise.resolve(0)), // Mock for count
    issueUniformItem: jest.fn(),
    createUniformItems: jest.fn(),
    getUniformItemDeficiencies: jest.fn(),
}));
jest.mock('../globalDataProvider', () => ({
    useGlobalData: jest.fn(() => ({
        sizelists: sizeLists,
        userRole: global.__ROLE__ ?? AuthRole.admin,
    })),
}));
jest.mock('@/dataFetcher/uniform', () => ({
    useUniformItemHistory: jest.fn(() => ({
        history: mockUniformHistory,
    })),
}));

// ------------- OTHER MOCKS -------------
jest.mock("swr", () => ({
    mutate: jest.fn(async () => { }),
}));

jest.mock('react-toastify', () => ({
    toast: {
        error: jest.fn(),
        success: jest.fn(),
    },
}));

export const mockDeficiencyTypeList = [
    {
        id: 'de3860d4-c88e-4a7c-be4c-7e832eda31d4',
        name: 'Test Type',
        dependent: 'uniform',
    },
    {
        id: '61d28738-4757-4e98-bfbf-a82f8404a74a',
        name: 'Another One',
        dependent: 'uniform',
    },
];
export const mockDeficiencyList: Deficiency[] = [
    {
        id: '80c36a66-81e9-4c9c-8d56-a095a6e9e28b',
        comment: 'Test comment',
        description: 'Test description',
        dateCreated: new Date('2023-10-01T12:00:00Z'),
        dateUpdated: new Date('2023-10-01T12:00:00Z'),
        userCreated: 'user',
        userUpdated: 'user',
        typeId: mockDeficiencyTypeList[0].id,
        typeName: mockDeficiencyTypeList[0].name,
    },
    {
        id: 'a1b2c3d4-e5f6-7g8h-9i0j-k1l2m3n4o5p6',
        comment: 'Test comment 2',
        description: 'Test description 2',
        dateCreated: new Date('2023-10-02T12:00:00Z'),
        dateUpdated: new Date('2023-10-12T12:00:00Z'),
        userCreated: 'user',
        userUpdated: 'user1',
        typeId: mockDeficiencyTypeList[1].id,
        typeName: mockDeficiencyTypeList[1].name,
    },
    {
        id: 'b7c8d9e0-f1g2-h3i4-j5k6-l7m8n9o0p1q2',
        comment: 'Test comment 3',
        description: 'Test description 3',
        dateCreated: new Date('2023-10-03T12:00:00Z'),
        dateUpdated: new Date('2023-10-12T12:00:00Z'),
        dateResolved: new Date('2023-10-17T12:00:00Z'),
        userCreated: 'user',
        userUpdated: 'user1',
        userResolved: 'user2',
        typeId: mockDeficiencyTypeList[0].id,
        typeName: mockDeficiencyTypeList[0].name,
    },
];
export const mockUniform = {
    id: "c227ac23-93d4-42b5-be2e-956ea35c2db9",
    number: 2501,
    generation: generationLists[0][1],
    size: sizeLists[0].uniformSizes[0],
    comment: "Test comment",
    active: true,
    type: {
        id: typeList[0].id,
        name: typeList[0].name,
    },
};
export const mockUniformHistory = [
    {
        dateIssued: new Date('2023-10-01T12:00:00Z'),
        dateReturned: null,
        cadet: {
            firstname: 'John',
            lastname: 'Doe',
            recdelete: false,
        },
    },
    {
        dateIssued: new Date('2023-10-02T12:00:00Z'),
        dateReturned: new Date('2023-10-03T12:00:00Z'),
        cadet: {
            firstname: 'Jane',
            lastname: 'Smith',
            recdelete: true,
        },
    },
];