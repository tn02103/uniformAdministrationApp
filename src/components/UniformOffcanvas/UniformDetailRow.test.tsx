import "./UniformOffcanvasJestHelper";

import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { generationLists, sizeLists, typeList } from "../../../tests/_jestConfig/staticMockData";
import { UniformDetailRow, UniformDetailRowProps } from "./UniformDetailRow";
import { mockUniform } from "./UniformOffcanvasJestHelper";

describe('UniformDetailRow', () => {
    const defaultProps: UniformDetailRowProps = {
        uniform: mockUniform,
        uniformType: typeList[0],
        editable: false,
        setEditable: jest.fn(),
        onSave: jest.fn(),
    };

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render correctly', () => {
        const { container } = render(<UniformDetailRow {...defaultProps} />);
        expect(container).toMatchSnapshot();
    });

    it('should render correctly when editable', async () => {
        const { container, rerender } = render(<UniformDetailRow {...defaultProps} />);
        rerender(<UniformDetailRow {...defaultProps} editable />);

        const generationSelect = screen.getByRole('combobox', { name: /generation/i });
        const sizeSelect = screen.getByRole('combobox', { name: /size/i });
        const commentInput = screen.getByRole('textbox', { name: /comment/i });
        const activeCheckbox = screen.getByRole('switch', { name: /status/i });

        expect(generationSelect).toBeInTheDocument();
        expect(sizeSelect).toBeInTheDocument();
        expect(commentInput).toBeInTheDocument();
        expect(activeCheckbox).toBeInTheDocument();

        expect(generationSelect).toHaveValue(mockUniform.generation.id);
        expect(sizeSelect).toHaveValue(mockUniform.size.id);
        expect(commentInput).toHaveValue(mockUniform.comment);
        expect(activeCheckbox).toBeChecked();

        expect(generationSelect).toBeEnabled();
        expect(sizeSelect).toBeEnabled();
        expect(commentInput).toBeEnabled();
        expect(activeCheckbox).toBeEnabled();

        expect(container).toMatchSnapshot();
    });


    it('should show size if not in sizelist', async () => {
        const uniform = {
            ...mockUniform,
            size: {
                id: 'nonexistent-size',
                name: 'Nonexistent Size',
            },
        };

        const { rerender } = render(<UniformDetailRow {...defaultProps} uniform={uniform} />);

        expect(screen.getByText('Nonexistent Size')).toBeInTheDocument();

        rerender(<UniformDetailRow {...defaultProps} uniform={uniform} editable />);
        const sizeSelect = screen.getByRole('combobox', { name: /size/i });
        expect(sizeSelect).toBeInTheDocument();
        expect(sizeSelect).toHaveValue('');
        expect(sizeSelect).toHaveTextContent('common.error.pleaseSelect');
    });

    it('should show correct size options', async () => {
        const user = userEvent.setup();
        const { rerender } = render(<UniformDetailRow {...defaultProps} />);
        rerender(<UniformDetailRow {...defaultProps} editable />);

        const sizeSelect = screen.getByRole('combobox', { name: /size/i });
        const generationSelect = screen.getByRole('combobox', { name: /generation/i });
        expect(sizeSelect).toBeInTheDocument();

        // initial state with sizelist
        const options = sizeSelect.querySelectorAll('option');
        expect(options.length).toBe(sizeLists[0].uniformSizes.length + 1); // +1 for the default "please select" option
        sizeLists[0].uniformSizes.forEach(size => {
            expect(screen.getByText(size.name)).toBeInTheDocument();
        });

        // different generation with sizelist
        await user.selectOptions(generationSelect, generationLists[0][2].id);
        const seccondOptions = sizeSelect.querySelectorAll('option');
        expect(seccondOptions.length).toBe(sizeLists[1].uniformSizes.length + 1); // +1 for the default "please select" option
        sizeLists[1].uniformSizes.forEach(size => {
            expect(screen.getByText(size.name)).toBeInTheDocument();
        });

        // different generation without sizelist
        await user.selectOptions(generationSelect, generationLists[0][3].id);
        const thirdOptions = sizeSelect.querySelectorAll('option');
        expect(thirdOptions.length).toBe(sizeLists[2].uniformSizes.length + 1); // +1 for the default "please select" option
        sizeLists[2].uniformSizes.forEach(size => {
            expect(screen.getByText(size.name)).toBeInTheDocument();
        });
    });

    it('should only keep size if in new sizelist', async () => {
        const user = userEvent.setup();
        const { rerender } = render(<UniformDetailRow {...defaultProps} />);
        rerender(<UniformDetailRow {...defaultProps} editable />);

        const sizeSelect = screen.getByRole('combobox', { name: /size/i });
        const generationSelect = screen.getByRole('combobox', { name: /generation/i });

        // initial state with sizelist
        expect(sizeSelect).toHaveValue(mockUniform.size.id);

        // different generation with sizelist
        await user.selectOptions(generationSelect, generationLists[0][2].id);
        expect(sizeSelect).toHaveValue(mockUniform.size.id);

        // different generation without sizelist
        await user.selectOptions(generationSelect, generationLists[0][3].id);
        expect(sizeSelect).toHaveValue('');
        expect(sizeSelect).toHaveTextContent('common.error.pleaseSelect');
    });

    it('should hide sizeField if !usingSizes', () => {
        render(<UniformDetailRow {...defaultProps} uniformType={typeList[1]} />);

        expect(screen.queryByLabelText('common.uniform.size')).not.toBeInTheDocument();
        expect(screen.queryByText(defaultProps.uniform.size?.name!)).not.toBeInTheDocument();
    });

    it('should hide generationField if !usingGenerations', () => {
        render(<UniformDetailRow {...defaultProps} uniformType={typeList[2]} />);

        expect(screen.queryByLabelText('common.uniform.generation.label')).not.toBeInTheDocument();
        expect(screen.queryByText(defaultProps.uniform.generation?.name!)).not.toBeInTheDocument();
    });

    describe('handle update to uniform', () => {
        const newUniform = {
            ...mockUniform,
            generation: generationLists[0][0],
            size: sizeLists[0].uniformSizes[1],
            comment: 'New comment',
            active: false,
        };
        it('should reset form when uniform prop changes', () => {
            const { rerender } = render(<UniformDetailRow {...defaultProps} />);
            const generationSelect = screen.getByLabelText('common.uniform.generation.label');
            const sizeSelect = screen.getByLabelText('common.uniform.size');
            const commentInput = screen.getByLabelText('common.comment');
            const activeCheckbox = screen.getByLabelText('common.status');

            expect(generationSelect).toHaveTextContent(mockUniform.generation.name);
            expect(sizeSelect).toHaveTextContent(mockUniform.size.name);
            expect(commentInput).toHaveTextContent(mockUniform.comment);
            expect(activeCheckbox).toHaveTextContent('common.uniform.active.true');

            rerender(<UniformDetailRow {...defaultProps} uniform={newUniform} />);

            expect(generationSelect).toHaveTextContent(newUniform.generation.name);
            expect(sizeSelect).toHaveTextContent(newUniform.size.name);
            expect(commentInput).toHaveTextContent(newUniform.comment);
            expect(activeCheckbox).toHaveTextContent('common.uniform.active.false');
        });

        it('should not reset if editable', () => {
            const { rerender } = render(<UniformDetailRow {...defaultProps} />);
            rerender(<UniformDetailRow {...defaultProps} editable />);

            const generationSelect = screen.getByLabelText('common.uniform.generation.label');
            const sizeSelect = screen.getByLabelText('common.uniform.size');
            const commentInput = screen.getByLabelText('common.comment');
            const activeCheckbox = screen.getByLabelText('common.status');

            expect(generationSelect).toHaveTextContent(mockUniform.generation.name);
            expect(sizeSelect).toHaveTextContent(mockUniform.size.name);
            expect(commentInput).toHaveTextContent(mockUniform.comment);
            expect(activeCheckbox).toBeChecked();

            rerender(<UniformDetailRow {...defaultProps} uniform={newUniform} editable />);

            expect(generationSelect).toHaveTextContent(mockUniform.generation.name);
            expect(sizeSelect).toHaveTextContent(mockUniform.size.name);
            expect(commentInput).toHaveTextContent(mockUniform.comment);
            expect(activeCheckbox).toBeChecked();
        });
    });
    describe('edit uniform', () => {
        it('should call updateUniformItem', async () => {
            const { updateUniformItem } = jest.requireMock('@/dal/uniform/item/_index');
            const user = userEvent.setup();
            const { rerender } = render(<UniformDetailRow {...defaultProps} />);
            rerender(<UniformDetailRow {...defaultProps} editable />);

            const generationSelect = screen.getByRole('combobox', { name: /generation/i });
            const sizeSelect = screen.getByRole('combobox', { name: /size/i });
            const commentInput = screen.getByRole('textbox', { name: /comment/i });
            const activeCheckbox = screen.getByRole('switch', { name: /status/i });
            const saveButton = screen.getByRole('button', { name: /save/i });

            await user.selectOptions(generationSelect, generationLists[0][0].id);
            await user.selectOptions(sizeSelect, sizeLists[0].uniformSizes[1].id);
            await user.clear(commentInput);
            await user.type(commentInput, 'New comment');
            await user.click(activeCheckbox);

            expect(generationSelect).toHaveValue(generationLists[0][0].id);
            expect(sizeSelect).toHaveValue(sizeLists[0].uniformSizes[1].id);
            expect(commentInput).toHaveValue('New comment');
            expect(activeCheckbox).not.toBeChecked();

            await user.click(saveButton);
            expect(defaultProps.onSave).toHaveBeenCalledTimes(1);
            expect(defaultProps.onSave).toHaveBeenCalledWith('Saved item');
            expect(defaultProps.setEditable).toHaveBeenCalledTimes(1);
            expect(defaultProps.setEditable).toHaveBeenCalledWith(false);

            expect(updateUniformItem).toHaveBeenCalledTimes(1);
            expect(updateUniformItem).toHaveBeenCalledWith({
                number: mockUniform.number,
                id: mockUniform.id,
                generation: generationLists[0][0].id,
                size: sizeLists[0].uniformSizes[1].id,
                comment: 'New comment',
                active: false,
            });
        });

        it('should reset form on cancel', async () => {
            const user = userEvent.setup();

            const { rerender } = render(<UniformDetailRow {...defaultProps} />);
            rerender(<UniformDetailRow {...defaultProps} editable />);

            const generationSelect = screen.getByRole('combobox', { name: /generation/i });
            const sizeSelect = screen.getByRole('combobox', { name: /size/i });
            const commentInput = screen.getByRole('textbox', { name: /comment/i });
            const activeCheckbox = screen.getByRole('switch', { name: /status/i });
            const cancelButton = screen.getByRole('button', { name: /cancel/i });

            await user.selectOptions(generationSelect, generationLists[0][0].id);
            await user.selectOptions(sizeSelect, sizeLists[0].uniformSizes[1].id);
            await user.clear(commentInput);
            await user.type(commentInput, 'New comment');
            await user.click(activeCheckbox);

            expect(generationSelect).toHaveValue(generationLists[0][0].id);
            expect(sizeSelect).toHaveValue(sizeLists[0].uniformSizes[1].id);
            expect(commentInput).toHaveValue('New comment');
            expect(activeCheckbox).not.toBeChecked();

            await user.click(cancelButton);
            expect(defaultProps.setEditable).toHaveBeenCalledTimes(1);
            expect(defaultProps.setEditable).toHaveBeenCalledWith(false);

            expect(generationSelect).toHaveValue(mockUniform.generation.id);
            expect(sizeSelect).toHaveValue(mockUniform.size.id);
            expect(commentInput).toHaveValue(mockUniform.comment);
            expect(activeCheckbox).toBeChecked();
        });

        it('should not allow null value for generation', async () => {
            const props = {
                ...defaultProps,
                uniform: {
                    ...mockUniform,
                    generation: null,
                }
            }
            const user = userEvent.setup();
            const { rerender } = render(<UniformDetailRow {...props} />);
            rerender(<UniformDetailRow {...props} editable />);

            const generationSelect = screen.getByRole('combobox', { name: /generation/i });
            const saveButton = screen.getByRole('button', { name: /save/i });

            expect(generationSelect).toHaveValue('');

            await user.click(saveButton);
            expect(defaultProps.onSave).not.toHaveBeenCalled();
            expect(generationSelect).toBeInvalid();
            expect(generationSelect).toHaveTextContent('common.error.pleaseSelect');
        });

        it('should not allow null value for size', async () => {
            const props = {
                ...defaultProps,
                uniform: {
                    ...mockUniform,
                    size: null,
                }
            }
            const user = userEvent.setup();
            const { rerender } = render(<UniformDetailRow {...props} />);
            rerender(<UniformDetailRow {...props} editable />);

            const sizeSelect = screen.getByRole('combobox', { name: /size/i });
            const saveButton = screen.getByRole('button', { name: /save/i });

            expect(sizeSelect).toHaveValue('');

            await user.click(saveButton);
            expect(defaultProps.onSave).not.toHaveBeenCalled();
            expect(sizeSelect).toBeInvalid();
            expect(sizeSelect).toHaveTextContent('common.error.pleaseSelect');
        });
        it('should allow null values if not using sizes or generations', async () => {
            const props = {
                ...defaultProps,
                uniformType: typeList[3],
                uniform: {
                    ...mockUniform,
                    generation: null,
                    size: null,
                }
            }
            const user = userEvent.setup();
            const { rerender } = render(<UniformDetailRow {...props} />);
            rerender(<UniformDetailRow {...props} editable />);

            const generationSelect = screen.queryByRole('combobox', { name: /generation/i });
            const sizeSelect = screen.queryByRole('combobox', { name: /size/i });
            const saveButton = screen.getByRole('button', { name: /save/i });

            expect(generationSelect).toBeNull();
            expect(sizeSelect).toBeNull();

            await user.click(saveButton);
            expect(defaultProps.onSave).toHaveBeenCalledTimes(1);
        });

        it('should not allow size not in sizelist', async () => {
            const props = {
                ...defaultProps,
                uniform: {
                    ...mockUniform,
                    size: {
                        id: 'nonexistent-size',
                        name: 'Nonexistent Size',
                    },
                }
            }
            const user = userEvent.setup();
            const { rerender } = render(<UniformDetailRow {...props} />);
            rerender(<UniformDetailRow {...props} editable />);

            const sizeSelect = screen.getByRole('combobox', { name: /size/i });
            const saveButton = screen.getByRole('button', { name: /save/i });

            expect(sizeSelect).toHaveValue('');

            await user.click(saveButton);
            expect(defaultProps.onSave).not.toHaveBeenCalled();
            expect(sizeSelect).toBeInvalid();
            expect(sizeSelect).toHaveTextContent('common.error.pleaseSelect');
        });
    });
});
