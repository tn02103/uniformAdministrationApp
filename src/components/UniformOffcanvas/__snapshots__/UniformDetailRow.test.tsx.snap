// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniformDetailRow should render correctly 1`] = `
<div>
  <form
    autocomplete="off"
    class="mb-4"
    novalidate=""
  >
    <div
      class="mb-3 mt-2 justify-content-between row"
    >
      <div
        class="col-sm-auto col-12"
      >
        <div>
          <label
            class="fw-bold m-0  form-label"
            for="uniform_toggle-active"
          >
            common.status
          </label>
          <p
            aria-label="common.status"
            aria-readonly="true"
            class="mt-2"
          >
            common.uniform.state.active
          </p>
          <div
            aria-label="error message active"
            class="text-danger fs-7"
            data-testid="err_active"
            id="uniform_err_active"
            role="alert"
          />
        </div>
      </div>
      <div
        class="col-sm-auto col-7"
      >
        <div>
          <label
            class="fw-bold m-0  form-label"
            for="uniform_select-generation"
          >
            common.uniform.generation.label
          </label>
          <p
            aria-label="common.uniform.generation.label"
            aria-readonly="true"
            class="py-2 m-0"
          >
            Generation1-2
          </p>
          <div
            aria-label="error message generation"
            class="text-danger fs-7"
            data-testid="err_generation"
            id="uniform_err_generation"
            role="alert"
          />
        </div>
      </div>
      <div
        class="p-0 col-sm-auto col-5"
        style="min-width: 100px;"
      >
        <div>
          <label
            class="fw-bold m-0  form-label"
            for="uniform_select-size"
          >
            common.uniform.size
          </label>
          <p
            aria-label="common.uniform.size"
            aria-readonly="true"
            class="py-2 m-0"
          >
            0
          </p>
          <div
            aria-label="error message size"
            class="text-danger fs-7"
            data-testid="err_size"
            id="uniform_err_size"
            role="alert"
          />
        </div>
      </div>
      <div
        class="col-12"
      >
        <div>
          <label
            class="fw-bold m-0  form-label"
            for="uniform_input-comment"
          >
            common.comment
          </label>
          <textarea
            aria-invalid="false"
            class="form-control-plaintext"
            disabled=""
            id="uniform_input-comment"
            name="comment"
            rows="2"
            width="auto"
          >
            Test comment
          </textarea>
          <div
            aria-label="error message comment"
            class="text-danger fs-7"
            data-testid="err_comment"
            id="uniform_err_comment"
            role="alert"
          />
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`UniformDetailRow should render correctly when editable 1`] = `
<div>
  <form
    autocomplete="off"
    class="mb-4"
    novalidate=""
  >
    <div
      class="mb-3 mt-2 justify-content-between row"
    >
      <div
        class="col-sm-auto col-12"
      >
        <div>
          <label
            class="fw-bold m-0  form-label"
            for="uniform_toggle-active"
          >
            common.status
          </label>
          <div
            class="form-check form-switch px-0 my-auto"
          >
            <input
              aria-checked="true"
              aria-describedby="uniform_active_toggleText"
              aria-disabled="false"
              checked=""
              class="form-check-input ms-0 my-2"
              id="uniform_toggle-active"
              name="active"
              role="switch"
              style="cursor: pointer;"
              tabindex="0"
              type="checkbox"
              value="true"
            />
            <label
              class="form-check-label ms-2"
              id="uniform_active_toggleText"
            >
              common.uniform.state.active
            </label>
          </div>
          <div
            aria-label="error message active"
            class="text-danger fs-7"
            data-testid="err_active"
            id="uniform_err_active"
            role="alert"
          />
        </div>
      </div>
      <div
        class="col-sm-auto col-7"
      >
        <div>
          <label
            class="fw-bold m-0  form-label"
            for="uniform_select-generation"
          >
            common.uniform.generation.label
          </label>
          <select
            aria-invalid="false"
            class="text-truncate  form-select"
            id="uniform_select-generation"
            name="generation"
          >
            <option
              disabled=""
              value=""
            >
              common.error.pleaseSelect
            </option>
            <option
              value="8c0e873b-2b7d-4624-8a61-a192146f6587"
            >
              Generation1-1
            </option>
            <option
              value="0292fdd9-9d47-4470-86b6-107c6f8797e4"
            >
              Generation1-2
            </option>
            <option
              value="6649bcce-7ea2-4f8c-b5cc-85242f2f4dec"
            >
              Generation1-3
            </option>
            <option
              value="654d3e11-caee-410d-af04-427a052d37dd"
            >
              Generation1-4
            </option>
          </select>
          <div
            aria-label="error message generation"
            class="text-danger fs-7"
            data-testid="err_generation"
            id="uniform_err_generation"
            role="alert"
          />
        </div>
      </div>
      <div
        class="p-0 col-sm-auto col-5"
        style="min-width: 100px;"
      >
        <div>
          <label
            class="fw-bold m-0  form-label"
            for="uniform_select-size"
          >
            common.uniform.size
          </label>
          <select
            aria-invalid="false"
            class="text-truncate  form-select"
            id="uniform_select-size"
            name="size"
          >
            <option
              disabled=""
              value=""
            >
              common.error.pleaseSelect
            </option>
            <option
              value="bc27d3eb-20f5-46fa-9140-7bb85c89b5bc"
            >
              0
            </option>
            <option
              value="3e41ad07-23c6-44f6-b1eb-3f620a6811f8"
            >
              1
            </option>
            <option
              value="caef6dc5-9af5-461e-8025-551e9e240957"
            >
              2
            </option>
            <option
              value="51d811d5-e624-4c03-859b-3fc2c26d185e"
            >
              3
            </option>
            <option
              value="4ec167be-923b-4237-98de-fc62283f9c42"
            >
              4
            </option>
          </select>
          <div
            aria-label="error message size"
            class="text-danger fs-7"
            data-testid="err_size"
            id="uniform_err_size"
            role="alert"
          />
        </div>
      </div>
      <div
        class="col-12"
      >
        <div>
          <label
            class="fw-bold m-0  form-label"
            for="uniform_input-comment"
          >
            common.comment
          </label>
          <textarea
            aria-invalid="false"
            class="form-control"
            id="uniform_input-comment"
            name="comment"
            rows="2"
            width="auto"
          >
            Test comment
          </textarea>
          <div
            aria-label="error message comment"
            class="text-danger fs-7"
            data-testid="err_comment"
            id="uniform_err_comment"
            role="alert"
          />
        </div>
      </div>
    </div>
    <div
      class="justify-content-evenly row"
    >
      <div
        class="col-auto"
      >
        <button
          class="btn btn-outline-secondary"
          type="button"
        >
          common.actions.cancel
        </button>
      </div>
      <div
        class="col-auto"
      >
        <button
          class="btn btn-outline-primary"
          type="submit"
        >
          common.actions.save
        </button>
      </div>
    </div>
  </form>
</div>
`;
